In [2]:
import numpy as np
import matplotlib.pyplot as plt
import time 

In [3]:
def gradient_descent(gradient,init_,learn_rate, n_iter=50, tol=1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate*gradient(x)
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return round(x*1000)/1000


#Minimia of eqation x^2+3x+4
c = gradient_descent(gradient=lambda v: 2 * v + 3, init_=4.0, learn_rate=0.2)      

#Minima of eqaution x^4-3x^2+2x
d = gradient_descent(gradient=lambda v: 4*(v**3)-6*v+2 , init_=-1, n_iter = 100,learn_rate=0.02)

print(c,d)

-1.5 -1.366


Q1 a) i) The minima point is -1.5 <br>
Q1 a) ii) The minima point is -1.366 

In [4]:
#Gerating Random data
np.random.seed (0)
x_test = 2.5*np.random.randn(10000) + 1.5 
res = 1.5*np.random.randn(10000) 
y_test = 2 + 0.3*x_test + res

#### Q1 b)

In [6]:
#Gradeint funtion for linear regression
def gradient_linear(x,y,a,b):
    n = len(x)
    z = a*x+b 
    #Gradient for a
    a = (-2/n)*sum(x*(y-z))
    #Gradient for b
    b = (-2/n)*sum(y-z)
    return np.array([a,b])


def gradient_descent(gradient,init_,learn_rate,x_test,y_test, n_iter=50, tol=1e-6):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate*gradient(x_test,y_test,x[0],x[1])
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return np.round(x,4)

init = np.array([0,0]).astype(float)
print(gradient_descent(gradient_linear,init,0.01,x_test,y_test,n_iter=1000))

[0.2953 2.0232]


Q1 c) The Value of a = 0.2953 and Value of b = 2.023 

In [8]:
def stochastic_gradient_descent(gradient,init_,learn_rate,x_data,y_data, n_iter=50, batch=100,tol=1e-6):
    x = init_
    for _ in range(n_iter):
        #randomly select one batch
        select = np.random.randint(batch)
        batch_1 = x_data[select]
        batch_2 = y_data[select]  
        delta = -learn_rate*gradient(batch_1,batch_2,x[0],x[1])
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return np.round(x,3)

#The batch size of each batch is 250 
number_of_batch = 40
x_batch = np.split(x_test,number_of_batch)
y_batch = np.split(y_test,number_of_batch)

#Calculating values using mini batch stochastic gradient desent 
init = np.array([0,0]).astype(float)
print(stochastic_gradient_descent(gradient_linear,init,0.001,x_batch,y_batch,n_iter=4000,batch = number_of_batch))

[0.3   2.016]


Q1 d) Value of a ~ 0.29 and b ~ 2.01 for minibatch stochastic gradient descent  with batch-size of 250


In [10]:
#stores the output and size of each batch
size = []
output = []
#stores time required for each batch size
req_time = []

temp = len(x_test)
x_batch = np.split(x_test,temp)
y_batch = np.split(y_test,temp)

#Stochastic gradient descent
init = np.array([0,0]).astype(float)

start = time.time()
output.append(stochastic_gradient_descent(gradient_linear,init,0.001,x_batch,y_batch,n_iter=15000,batch = temp))
end = time.time()

req_time.append(end-start)
size.append(1)
print('SGD',end-start)

#Gradient Descent
init = np.array([0,0]).astype(float)
x_batch = np.split(x_test,1)    
y_batch = np.split(y_test,1)
start = time.time()
output.append(stochastic_gradient_descent(gradient_linear,init,0.001,x_batch,y_batch,n_iter=15000,batch = 1))
end = time.time()
print('GD',end-start)

SGD 0.05103135108947754
GD 18.241029024124146


Time for stochastic gradeint descent is samller than gradeint descent 

In [11]:
#calulating time and mse for each batch-size
for i in range(0,2000,50):
    init = np.array([0,0]).astype(float)
    try:
        number_of_batch = int(len(x_test)/i)
        x_batch = np.split(x_test,number_of_batch)    
        y_batch = np.split(y_test,number_of_batch)
    except:
        continue
    #print(len(x_batch))
    start = time.time()
    output.append(stochastic_gradient_descent(gradient_linear,init,0.001,x_batch,y_batch,n_iter=15000,batch = number_of_batch))
    size.append(i)
    end = time.time()
    req_time.append(end-start)

In [12]:
#For calculating MSE
def MSE(x,y,a,b):
    y_pred = x*a+b
    m = len(x)
    mse = (1/m)*(sum((y-y_pred)**2))
    return mse

min = 100
for i in range(0,len(size)):
    current = MSE(x_test,y_test,output[i][0],output[i][1])
    if min>current:
        min = current
        number = size[i]
    print('Batch size - ', size[i],'|Value of a , b -',output[i][0],output[i][1],'|MSE - ', current,'|Time - ',req_time[i])
print('\nBatch with minimum mse ',min,number)

Batch size -  1 |Value of a , b - 0.26 1.943 |MSE -  2.243653157691015 |Time -  0.05103135108947754
Batch size -  50 |Value of a , b - 0.295 2.023 |MSE -  2.2187221927113865 |Time -  0.7008380889892578
Batch size -  100 |Value of a , b - 0.296 2.03 |MSE -  2.2187831661516815 |Time -  0.9458296298980713
Batch size -  200 |Value of a , b - 0.288 2.021 |MSE -  2.2192146433600657 |Time -  1.415416955947876
Batch size -  250 |Value of a , b - 0.297 2.02 |MSE -  2.2187389555239108 |Time -  1.7471368312835693
Batch size -  400 |Value of a , b - 0.293 2.027 |MSE -  2.218753891450322 |Time -  2.4499435424804688
Batch size -  500 |Value of a , b - 0.295 2.022 |MSE -  2.218724695885996 |Time -  2.9453635215759277
Batch size -  600 |Value of a , b - 0.29 2.023 |MSE -  2.2189577673684893 |Time -  3.932192325592041
Batch size -  950 |Value of a , b - 0.294 2.024 |MSE -  2.218733059453963 |Time -  5.720537900924683
Batch size -  1000 |Value of a , b - 0.297 2.022 |MSE -  2.218739580500268 |Time -  5.

Q1 e) With respect to time Gradient Descent is slower than Stochastic Gradient Descent <br>
i.e. Time required - GD > Mini-batch SGD > SGD
<br> The GD is slower as it take all the data input for each iteration and then update the value of parameters whereas in SGD it take only one data and then update the value of paramenters and the computational complexity higher for GD and lower for SGD for each iteration<br>
In terms of MSE there is no large significant difference but for the optimal batch size we can choose the smaller batch size which gives us similar result in less time

###Q2 a)

$\text{Using Conditional Probabiltiy} $ <br><br>
$P(X|Y) = \frac{P(X\cap Y)}{P(X)} $ <br><br>
$P(X \cap Y) = P(X|Y)*P(X)$ <br><br>
$ \text{ Let us consider $X$ = Someone  Fever and $Y$ = Someone has Cold }$<br><br>
$P(\text{Someone has cold  $\cap$ Someone has Fever }) = P(\text{Someone has Fever}|\text{Someone has Cold})* P(\text{Someone has Cold})$<br>
<br>
$\text{From bayesian network we know that }$<br><br>
$P(X|Y) = 0.307 \text{  and  } P(X) = 0.02$<br><br>
$P(\text{Someone has cold  $\cap$ Someone has Fever }) =P(X \cap Y) = 0.307*0.02 = 0.00614 $

###Q2 b

$\text{To calulate $P$(Cold|Cough)}\\\text{First we need to calculate $P$(Cough)}$ <br><br>
$\text{To calulate $P$(Cough) we need to calculate $P$(Lung Disease) and $P$(Not Lung Disease)}$<br><br>
$\begin{equation}\begin{split}P(\text{Lung Disease}) & = P(\text{Lung Disease|Smokes}) * P(\text{Smokes}) + P(\text{Lung Disease|Not Smokes})*P(\text{Not Smokes})\\ & = 0.1009*0.2 + 0.001*0.8 \\ & = 0.02098\end{split}\end{equation}$ <br><br>

$\begin{equation}\begin{split}P(\text{Not Lung Disease}) & =1 - P(\text{Lung Disease})\\ & = 1 - 0.02098\\ & = 0.97902\end{split}\end{equation}$ 
<br><br>

$\begin{equation}\begin{split}P(\text{Cough}) & = P(\text{Cough | Cold $\cap$ Lung Disease }) * P(\text{Cold $\cap$ Lung Disease }) + P(\text{Cough | Cold $\cap$ Not Lung Disease }) * P(\text{Cold $\cap$ Not Lung Disease })+ P(\text{Cough | Not Cold $\cap$ Lung Disease }) * P(\text{Not Cold $\cap$ Lung Disease } + P(\text{Cough | Not Cold $\cap$ Not Lung Disease }) * P(\text{Not Cold $\cap$ Not Lung Disease }))
+\\ & =(0.7525* 0.02098* 0.02) + (0.505* 0.97902 *0.02) +(0.505* 0.02098 *0.98) +(0.01* 0.97902 * 0.98) \\ & = 0.030181249\end{split}\end{equation}$<br><br>

$\begin{equation}\begin{split}P(\text{Cough | Cold}) & = P(\text{Cough|Lung Diesease $\cap$ Cold}) * P(\text{Lung Disease}) + P(\text{Cough | Not Lung Disease $\cap$ Cold})*P(\text{Not Lung Disease})\\ & = 0.7525*0.02098 + 0.505*0.97902 \\ & = 0.51019225\end{split}\end{equation}$<br><br><br><br>
$\text{Using Bayes Rules}$<br><br>
$\begin{equation}\begin{split}P(\text{Cold | Cough}) & = \frac{P(\text{Cough | Cold})* P(\text{Cold})}{P(\text{Cough})}  \\& = \frac{0.51019225*0.02}{0.030181249} \\ & =0.3380857764\end{split}\end{equation}$

### Q3
<br>
$\text{In Multinomial distribution} \\ \displaystyle f(x_1,x_2,x_3,...,x_k |p_1,p_2,...,p_k) = \frac{n!}{\prod_{i=1}^{k}x_i}  \prod_{i=1}^{k}p_i^{x_i} \\
\text{where} \sum_{j=1}^{k}x_i = n \text{ and } \sum_{i=1}^{k}p_i = 1 $

<br>
$\text{Maximum Likelihood Estimation}$<br> 
$\text{The log likelihood is}\\
\begin{align}
l(\text{p}) = \log L(\text{p}) 
&= \log n! + \log \prod_{i=1}^k \frac{p_i^{x_i}}{x_i!} \\
&= \log n! + \sum_{i=1}^k \log \frac{p_i^{x_i}}{x_i!} \\
&= \log n! + \sum_{i=1}^k x_i \log p_i - \sum_{i=1}^k \log x_i!
\end{align}$<br>

$ \text{Constraint : }\sum_{j=1}^k p_j =1\\
\text{As we have constraint we will use Lagrange multipliers} \\
\begin{align}
l'(p,\lambda) &= l(p) + \lambda\bigg(1 - \sum_{i=1}^k p_i\bigg)
\end{align}$<br><br>
$\text{By posing derivative =0}\\
\begin{align}
\frac{\partial}{\partial p_i} l'(p,\lambda) 
= \frac{\partial}{\partial p_i} l(p)
+ \frac{\partial}{\partial p_i} \lambda\bigg(1 - \sum_{i=1}^k p_i\bigg) &= 0\\
 \frac{\partial}{\partial p_i} \sum_{i=1}^k x_i \log p_i
- \lambda \frac{\partial}{\partial p_i} \sum_{i=1}^k p_i &= 0 \\
\frac{x_i}{p_i}- \lambda  &= 0  \\
p_i &= \frac{x_i}{\lambda}  \quad \quad \quad (1)\\
\sum_{i=1}^k p_i &= \sum_{i=1}^k \frac{x_i}{\lambda} \\
\text{We know that  } \sum_{i=1}^k p_i & =1\\
1 &= \frac{1}{\lambda} \sum_{i=1}^k x_i \\
\lambda &= n \quad \quad \quad (2)\\
 \quad \quad \quad \text{ From 1 and 2 } \\
p_i &= \frac{x_i}{n}  \\ 
\text{p}& = \bigg(
\frac{x_1}{n},
...,
\frac{x_k}{n}
\bigg)
\end{align}$


